### PR TITLE
UrlClient refactor

### PR DESCRIPTION
**Pull request by madmaxoft**
(This PR is only to check the builds on the buildserver; the code has been offered to @12xx12 's PR, through which it should make it back to this repo)

### CODE REVIEW

**Comment by bearbin** [comment]
Good news, your build passed :)

**Comment by 12xx12** [comment]
I have time again on Friday. Take over my PR if you want to. This should be enabled for maintainers

**Comment by madmaxoft** [comment]
I can wait, no problem. I'd prefer another set of eyes (and brain) to have a look, so that I don't put some nonsense in.

**Comment by madmaxoft** [comment]
Okay, I polished this PR as a replacement to #5480 , squashing @12xx12 's commits into a single one, since it had a few dead-ends. This is now worth merging, preferably by rebase-merging all its commits intact.

**Comment by 12xx12** [comment]
So we disable cert checking again by default but enable users to supply their own certs.

And still use the cUrlClient for account checking

**Comment by madmaxoft** [comment]
Yes.
Basically, this adds an opt-in cert checking (which is an improvement from status quo), and uses `cUrlClient` for the Mojang API, with cert-checks.

**Review by 12xx12 — APPROVED** [review]

**Comment by madmaxoft** [comment]
Ouch, I forgot to update LuaAPI docs about the new `TrustedRootCAs` option. Will add.

**Comment by 12xx12** [comment]
I thought this was checked by the api checker?

**Comment by 12xx12** [comment]
Still the function deviates from the documented one

**Comment by madmaxoft** [comment]
Which one and how?

**Comment by madmaxoft** [comment]
Oh, Lua's `cTCPLink::StartTLSClient`, right, fixing that as well.
No, the checker cannot detect mismatch between the documented and real params, unfortunately.

**Comment by 12xx12** [comment]
The only thing i noticed, is that there is no note on how the new parameters work

And maybe we should improve the checker to include the params

File: Server/Plugins/APIDump/Classes/Network.lua
Comment (mathiascode) [inline Server/Plugins/APIDump/Classes/Network.lua:548]:
```suggestion
				Notes = "Starts a TLS handshake on the link, as a client side of the TLS. The Own___ parameters specify the client certificate and its corresponding private key and password; all three parameters are optional and no client certificate is presented to the remote peer if they are not used or all empty. Once the TLS handshake is started by this call, all incoming data is first decrypted before being sent to the OnReceivedData callback, and all outgoing data is queued until the TLS handshake completes, and then sent encrypted over the link. Returns true on success, nil and optional error message on immediate failure.<br/>The TrustedRootCAs is a string containing all certificates that should be trusted, in PEM format, concatenated and separated by a newline.<b>NOTE:</b> If TrustedRootCAs is empty or nil, the server's certificate will NOT be verified, which is UNSAFE!",
```

File: src/Protocol/Authenticator.cpp
Comment (mathiascode) [inline src/Protocol/Authenticator.cpp:78]:
```suggestion
		auto [IsSuccessful, ErrorMessage] = cUrlParser::Validate(m_Server);
		if (!IsSuccessful)
		{
			LOGWARNING("%s %d: Supplied invalid URL for configuration value [Authentication: Server]: \"%s\", using default! Error: %s", __FUNCTION__, __LINE__, m_Server.c_str(), ErrorMessage.c_str());
			m_Server = DEFAULT_AUTH_SERVER;
		}
	}

	{
		auto [IsSuccessful, ErrorMessage] = cUrlParser::Validate(m_Server);
		if (!IsSuccessful)
```

File: src/Protocol/Authenticator.cpp
Comment (mathiascode) [inline src/Protocol/Authenticator.cpp:187]:
```suggestion
	auto [IsSuccessful, Response] = cUrlClient::BlockingGet(m_Server + ActualAddress);
	if (!IsSuccessful)
```

File: src/Protocol/Authenticator.cpp
Comment (mathiascode) [inline src/Protocol/Authenticator.cpp:234]:
```suggestion
	auto [IsSuccessful, Response] = cUrlClient::BlockingGet(m_Server + ActualAddress);
	if (!IsSuccessful)
```

File: src/Protocol/MojangAPI.cpp
Comment (mathiascode) [inline src/Protocol/MojangAPI.cpp:557]:
```suggestion
		auto [IsSuccessful, Response] = cUrlClient::BlockingPost(m_NameToUUIDUrl, {}, std::move(RequestBody), MojangTrustedRootCAs::UrlClientOptions());
		if (!IsSuccessful)
```

File: src/Protocol/MojangAPI.cpp
Comment (mathiascode) [inline src/Protocol/MojangAPI.cpp:637]:
```suggestion
	auto [IsSuccessful, Response] = cUrlClient::BlockingGet(Url, {}, {}, MojangTrustedRootCAs::UrlClientOptions());
	if (!IsSuccessful)
```

**Comment by madmaxoft** [comment]
I let Cuberite from this PR run for longer now and I'm getting this error:
```
[e695214535086eb7|19:36:14] cMojangAPI::QueryNamesToUUIDs failed: Cannot parse received data (NameToUUID) to JSON: "* Line 1, Column 1
  Syntax error: value, object or array expected.
"
[e695214535086eb7|19:36:14] Response body:
00000000: 3C 21 44 4F 43 54 59 50 45 20 48 54 4D 4C 20 50    <!DOCTYPE HTML P
00000010: 55 42 4C 49 43 20 22 2D 2F 2F 57 33 43 2F 2F 44    UBLIC "-//W3C//D
00000020: 54 44 20 48 54 4D 4C 20 34 2E 30 31 2F 2F 45 4E    TD HTML 4.01//EN
00000030: 22 22 68 74 74 70 3A 2F 2F 77 77 77 2E 77 33 2E    ""http://www.w3.
00000040: 6F 72 67 2F 54 52 2F 68 74 6D 6C 34 2F 73 74 72    org/TR/html4/str
00000050: 69 63 74 2E 64 74 64 22 3E 0D 0A 3C 48 54 4D 4C    ict.dtd">..<HTML
00000060: 3E 3C 48 45 41 44 3E 3C 54 49 54 4C 45 3E 42 61    ><HEAD><TITLE>Ba
00000070: 64 20 52 65 71 75 65 73 74 3C 2F 54 49 54 4C 45    d Request</TITLE
00000080: 3E 0D 0A 3C 4D 45 54 41 20 48 54 54 50 2D 45 51    >..<META HTTP-EQ
00000090: 55 49 56 3D 22 43 6F 6E 74 65 6E 74 2D 54 79 70    UIV="Content-Typ
000000a0: 65 22 20 43 6F 6E 74 65 6E 74 3D 22 74 65 78 74    e" Content="text
000000b0: 2F 68 74 6D 6C 3B 20 63 68 61 72 73 65 74 3D 75    /html; charset=u
000000c0: 73 2D 61 73 63 69 69 22 3E 3C 2F 48 45 41 44 3E    s-ascii"></HEAD>
000000d0: 0D 0A 3C 42 4F 44 59 3E 3C 68 32 3E 42 61 64 20    ..<BODY><h2>Bad
000000e0: 52 65 71 75 65 73 74 20 2D 20 49 6E 76 61 6C 69    Request - Invali
000000f0: 64 20 43 6F 6E 74 65 6E 74 20 4C 65 6E 67 74 68    d Content Length
00000100: 3C 2F 68 32 3E 0D 0A 3C 68 72 3E 3C 70 3E 48 54    </h2>..<hr><p>HT
00000110: 54 50 20 45 72 72 6F 72 20 34 30 30 2E 20 54 68    TP Error 400. Th
00000120: 65 72 65 20 69 73 20 61 6E 20 69 6E 76 61 6C 69    ere is an invali
00000130: 64 20 63 6F 6E 74 65 6E 74 20 6C 65 6E 67 74 68    d content length
00000140: 20 6F 72 20 63 68 75 6E 6B 20 6C 65 6E 67 74 68     or chunk length
00000150: 20 69 6E 20 74 68 65 20 72 65 71 75 65 73 74 2E     in the request.
00000160: 3C 2F 70 3E 0D 0A 3C 2F 42 4F 44 59 3E 3C 2F 48    </p>..</BODY></H
00000170: 54 4D 4C 3E 0D 0A                                  TML>..
```
Something is still wrong.

**Comment by NiLSPACE** [comment]
Looks like it's because it's sending the entire body as the `Content-Length` body:
https://github.com/cuberite/cuberite/blob/c9522fb740200ccef6230cec452c48efb31e5394/src/HTTP/UrlClient.cpp#L292

**Comment by NiLSPACE** [comment]
I'm surprised the UrlClient tests didn't fail. It seems every request the UrlClient makes in master should be malformed right now.

**Comment by madmaxoft** [comment]
The test only check that *something* is received, be it real content or an error message.
I guess the `cUrlClient` methods should indicate failure if an http error code is returned from the server, but that's for another PR. Let's get this merged already.

**Review by NiLSPACE — APPROVED** [review]

**Pull request merged**
